src/QsCompiler/Transformations/TypeResolutionSummary.cs

```diff
@@ -0,0 +1,390 @@
+            }
+
+            return combinedBuilder.ToImmutable();
+        }
+
+        /// <summary>
+        /// Combines the resolution dictionaries in the summary into one resolution dictionary containing
+        /// the resolutions for all the type parameters found.
+        /// Updates the summary with the constructed dictionary. Updates the summary flags accordingly.
+        /// </summary>
+        private void CombineTypeResolutions()
+        {
+            var combinedBuilder = ImmutableDictionary.CreateBuilder<TypeParameterName, ResolvedType>();
+
+            foreach (var resolutionDictionary in this.ResolutionsInSummary)
+            {
+                TypeParameterResolutions resolvedDictionary;
+                resolvedDictionary = this.CombineTypeResolutionDictionary(resolutionDictionary);
+
+                // Contains a lookup of all the keys in the combined resolutions whose value needs to be updated
+                // if a certain type parameter is resolved by the currently processed dictionary.
+                var mayBeReplaced = GetReplaceable(combinedBuilder);
+
+                // Do any replacements for type parameters that may be replaced with values in the current dictionary.
+                // This needs to be done first to cover an edge case.
+                foreach (var (typeParam, paramRes) in resolvedDictionary.Where(entry => mayBeReplaced.Contains(entry.Key)))
+                {
+                    this.UpdatedReplaceableResolutions(mayBeReplaced, combinedBuilder, typeParam, paramRes);
+                }
+
+                // Validate and add each resolution to the result.
+                foreach (var (typeParam, paramRes) in resolvedDictionary)
+                {
+                    // Check that we are not constricting a type parameter to another type parameter of the same callable.
+                    this.UpdateConstrictionFlag(typeParam, paramRes);
+
+                    // Check that there is no conflicting resolution already defined.
+                    if (!this.SummarizesOverConflictingResolution)
+                    {
+                        this.SummarizesOverConflictingResolution = combinedBuilder.TryGetValue(typeParam, out var current)
+                            && !current.Equals(paramRes) && !IsSelfResolution(typeParam, current);
+                    }
+
+                    // Add the resolution to the current dictionary.
+                    combinedBuilder[typeParam] = paramRes;
+                }
+            }
+
+            this.CombinedTypeParameterResolutions = combinedBuilder.ToImmutable();
+        }
+
+        // Nested Classes
+
+        /// <summary>
+        /// Walker that collects all of the type parameter references for a given ResolvedType
+        /// and returns them as a HashSet.
+        /// </summary>
+        private class GetTypeParameters : TypeTransformation<GetTypeParameters.TransformationState>
```

Comment:
I missed this in the last PR but since this is a class, maybe a name like `TypeParametersGetter` might be a name that conveys that its is a class?

Reply:
The standard is for transformation classes to have verbs for names. For examples you can look at the [Basic Transformations]( https://github.com/microsoft/qsharp-compiler/blob/master/src/QsCompiler/Transformations/BasicTransformations.cs): `GetSourceFiles`, `FilterBySourceFile`, `SelectByFoldingOverExpressions`, `SelectByAnyContainedExpression`, `SelectByAllContainedExpressions`, and `FoldOverExpressions`.